dimagi/commcare-hq
# Maintenance alert scheduling

## Product Description
Adds start time, end time, and timezone options to maintenance alerts. Activated alerts will be shown between the start time and end time, taking into account the provided timezone. Alerts with no schedule will still be shown immediately on activation.
<img width="739" alt="image" src="https://user-images.githubusercontent.com/100609770/207167246-d4b54d9a-abe5-4d53-b120-74905761717b.png">


Alerts will be displayed on the admin `/alerts/` page showing their start and end times in the provided timezone.
<img width="745" alt="image" src="https://user-images.githubusercontent.com/100609770/207166039-6444198b-ccef-4855-bc4f-bdcfc42d6007.png">


## Technical Summary
[USH-2703](https://dimagi-dev.atlassian.net/browse/USH-2703)
Adds fields for `start_time`, `end_time`, and `timezone` to the `MaintenanceAlert` model. The `get_active_alerts` method is modified to get alerts that are active and: have no schedule, or where it is currently past the alert's start time but before its end time. This method has its cache timeout decreased to 1 minute from 60, to regularly check for alerts that should be shown or hidden based on schedule.

Makes changes to the `MaintenanceAlertsView` to convert user-entered start and end datetimes to UTC using existing utilities before saving to model. Similarly converts stored UTC datetimes based on the saved timezone, before displaying to user. Consolidates the `activate_alert` and `deactivate_alert` functions under the view and removes these as separate urls.

Adds datetime pickers and a select2 element to the html template used for the `/alerts/` page as well as table columns to display timing. Uses browser validations to check that alert text is not blank, and end time (if present) is after start time (if present).

## Safety Assurance

### Safety story
`MaintenanceAlert` migration handles existing objects by allowing blank start and end times on model, setting timezone default to UTC. Timezone entry is populated from a `pytz` list of common timezones and is not a text entry. 

### Automated test coverage
Added test coverage for MaintenanceAlert model and MaintenanceAlertsView.

### QA Plan
Will send to QA to ensure new ui elements and scheduled alerts work predictably and without error.

### Migrations
- [X] The migrations in this code can be safely applied first independently of the code

### Rollback instructions




### Labels & Review
